test/auth_SUITE.erl

```diff
@@ -0,0 +1,287 @@
+           ok = gen_rpc_test_helper:start_slave(Driver),
+           ?assertMatch({badrpc, invalid_cookie}, gen_rpc:call(?SLAVE, ?MODULE, canary, []))
+       end,
+       [ fun ?MODULE:prop_canary/1
+       ]).
+
+%% In this testcase we don't test auth, but the rest of the gen_rpc library.
+%%
+%% We mock authentication to always fail and verify that it prevents access.
+t_auth_client_fail(Config) ->
+    Driver = gen_rpc_test_helper:get_driver_from_config(Config),
+    ?check_trace(
+       #{timetrap => 5000},
+       begin
+           meck:new(gen_rpc_auth, [passthrough]),
+           meck:expect(gen_rpc_auth, authenticate_client,
+                       fun(_Driver, _Socket, _Peer) ->
+                               {error, {badrpc, invalid_cookie}}
+                       end),
+           ok = gen_rpc_test_helper:start_master(Driver),
+           ok = gen_rpc_test_helper:start_slave(Driver),
+           Node = node(),
+           ?assertNotMatch(canary_is_dead,
+                           erpc:call(?SLAVE,
+                                     fun() ->
+                                             gen_rpc:call(Node, ?MODULE, canary, [])
+                                     end))
+       end,
+       [ fun ?MODULE:prop_canary/1
+       ]).
+
+%% The client has invalid cookie:
+t_challenge_response_invalid_cookie_client(Config) ->
+    Driver = gen_rpc_test_helper:get_driver_from_config(Config),
+    ?check_trace(
+       #{timetrap => 5000},
+       try
+           application:set_env(?APP, secret_cookie, <<"wrong">>),
+           ok = gen_rpc_test_helper:start_master(Driver),
+           ok = gen_rpc_test_helper:start_slave(Driver),
+           ?assertMatch({badrpc, invalid_cookie}, gen_rpc:call(?SLAVE, ?MODULE, canary, []))
+       after
+           application:unset_env(?APP, secret_cookie)
+       end,
+       [ fun ?MODULE:prop_canary/1
+       , fun ?MODULE:prop_no_fallback/1
+       ]).
+
+%% The server has invalid cookie:
+t_challenge_response_invalid_cookie_server(Config) ->
+    Driver = gen_rpc_test_helper:get_driver_from_config(Config),
+    ?check_trace(
+       #{timetrap => 5000},
+       begin
+           ok = gen_rpc_test_helper:start_master(Driver),
+           ok = gen_rpc_test_helper:start_slave(Driver),
+           erpc:call(?SLAVE,
+                     fun() ->
+                             application:set_env(?APP, secret_cookie, <<"wrong">>)
+                     end),
+           ?assertMatch({badrpc, invalid_cookie},
+                        gen_rpc:call(?SLAVE, ?MODULE, canary, []))
+       end,
+       [ fun ?MODULE:prop_canary/1
+       , fun ?MODULE:prop_no_fallback/1
+       ]).
+
+%% Compatibility (happy case)
+t_compat_old_server_ok(Config) ->
+    Driver = gen_rpc_test_helper:get_driver_from_config(Config),
+    application:set_env(?APP, insecure_auth_fallback_allowed, true),
+    ?check_trace(
+       #{timetrap => 5000},
+       begin
+           ok = gen_rpc_test_helper:start_master(Driver),
```

Comment:
In another test I send a gen_rpc call to the master from the slave which is equivalent. Changing the running code on the master is more challenging since a cleanup procedure would be needed.